osd: refactoring tier_flush()

myoungwon opened the pull request:
The objective of this PR is to refactor tier_flush() as described in TODO as below.

-   TODO: "flush" for a manifest object means re-running the CDC algorithm on the portions of the object that are not currently dedup'd (not in the manifest chunk_map) and re-deduping the resulting chunks.  Adding support for that operation here is future work.

Signed-off-by: Myoungwon Oh <myoungwon.oh@samsung.com>

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

myoungwon commented:
@athanatos Any comments?

athanatos commented:
I will try to review this next week.

myoungwon commented:
@athanatos Ping. Can you take a look?

athanatos commented:
Apologies, I have been busy with other things.  I'll try to look at it tomorrow or Monday.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9978,264 @@ struct C_Flush : public Context {
+  }
+  while (cur_off < oi.size && cur_off < max_window_size) {
+    bufferlist chunk_data;
+    // TODO: Do we need to support EC?
```
I think it's ok for now not to.  Just make sure to return an error code and add a test validating the error code behavior.

athanatos replied:
Probably the mon shouldn't allow an ec pool to be set as the top dedup pool?

athanatos replied:
That is, I think the pool settings added here should be invalid on an EC pool.

athanatos replied:
You left this as a TODO with a question.  Please change the comment to explain your current choice and reasons so that readers later won't have to guess.

athanatos commented on src/mon/OSDMonitor.cc:
```diff
@@ -8648,6 +8661,39 @@ int OSDMonitor::prepare_command_pool_set(const cmdmap_t& cmdmap,
+      }
+      const pg_pool_t *tp = osdmap.get_pg_pool(lowtierpool_id);
+      ceph_assert(tp);
+      if (p.is_erasure()) {
```
(This is not a comment on this PR, just a general question I have)  Why we are disallowing erasure for a base pool?  Is it because the class operations for refcounting don't work at the moment or something?

myoungwon replied:
The reason why I disallow erasure code pool here is that the EC pool does not support objects_read_sync(). Therefore, I should change the current implementation to make EC pool compatible. As s result, I think this can be future work. 
Also, without this, we can use a combination of a base tier (replicated) and a low tier (replicated or EC pool).

athanatos replied:
Got it.

athanatos commented on src/mon/OSDMonitor.cc:
```diff
@@ -8648,6 +8661,39 @@ int OSDMonitor::prepare_command_pool_set(const cmdmap_t& cmdmap,
+	return -ENOTSUP;
+      }
+      n = lowtierpool_id;
+      interr.clear(); 
```
I think you need to actually check interr here.  If I'm understanding the code correctly, val is only valid if interr.empty(), otherwise interr indicates why val couldn't be parsed.

athanatos replied:
I think I'm not following what happens here.  Rereading it, it looks like interr should not be relevant here as the param is actually a string, so it should always be empty?  You've added a check that returns an error if interr is empty.  I don't understand why it would normally be non-empty.

myoungwon replied:
Hm.. Input val is string (e.g., pool_name), so interr is true and interr.empty() is false here. But, the actual value supposed to store is pool_opts_t::INT(pool_name --> pool_id). As a result, if I don't clear interr, -EINVAL will be returned.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9978,264 @@ struct C_Flush : public Context {
+  
+  C_SetDedupChunks(PrimaryLogPG *p, hobject_t o, epoch_t lpr)
+    : pg(p), oid(o), last_peering_reset(lpr),
+      tid(0)
```
Add offset as an argument and initialize here.  I see only one user which sets offset immediately after calling the constructor, so this doesn't appear to be a problem.  It's basically never ok to leave a struct/class member uninitialized after the constructor.

athanatos replied:
You've initialized it to zero, but I want you to *pass in* the offset and initialize it *to the correct value*.  You use this in exactly one place:

```
      C_SetDedupChunks *fin = new C_SetDedupChunks(this, soid, get_last_peering_reset());
      fin->offset = p.first;
```

and then immediately reset offset to p.first.  Pass p.first into the constructor.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9978,264 @@ struct C_Flush : public Context {
+    }
+
+    it->second->results[offset] = r;
+    if (last_peering_reset == pg->get_last_peering_reset() &&
```
If last_peering_reset != pg->get_last_peering_reset(), then pg->manifest_ops must have been already cleared as part of the peering reset.  Do this check first immediately after taking the lock -- you should simply drop the lock and return in that case without checking any other pg state.

athanatos replied:
See cancel_manifest_ops.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9978,264 @@ struct C_Flush : public Context {
+  return target;
+}
+
+void PrimaryLogPG::finish_set_dedup(hobject_t oid, map<hobject_t, pair<uint64_t, uint64_t>>& chunks,
```
You are passing in chunks here -- a reference to a member of a ManifestOpRef ...

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9978,264 @@ struct C_Flush : public Context {
+{
+  dout(10) << __func__ << " " << oid << " tid " << tid
+	   << " " << cpp_strerror(r) << dendl;
+  map<hobject_t,ManifestOpRef>::iterator p = manifest_ops.find(oid);
```
... which you look up again here.  Why not simply pass in the iterator you must already have looked up to get chunks instead of chunks?  You've already validated that it exists.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9978,264 @@ struct C_Flush : public Context {
+      // if any failure occurs, put a mark on the results to recognize the failure
+      it->second->results[0] = r;
+      if (last_peering_reset != pg->get_last_peering_reset()) 
+	it->second->results[0] = -EINVAL;
```
If num_chunks == results.size(), you need to call into finish_set_dedup as well, I think.  finish_set_dedup then neeeds to check results and return an error if any were unsuccessful.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9978,264 @@ struct C_Flush : public Context {
+    manifest_ops[soid]->op = op;
+  } 
+
+  return -EINPROGRESS;
```
This is possible if mop->tids.size() == 0.  Does that make sense?

athanatos replied:
Is there a test case for the mop->tids.size() == 0 case?  Is it even possible?

athanatos replied:
I guess it happens if all regions are already dedup'd?

athanatos replied:
Found the test, nice.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9978,264 @@ struct C_Flush : public Context {
+  const hobject_t& soid = oi.soid;
+
+  if (!obc->is_blocked()) {
+    return -EINVAL;
```
I think you need to add more detail here.  If you want to disallow non-blocking writes for dedup objects, I think you probably want to reject the request much earlier -- probably in do_osd_ops.

athanatos replied:
Take a look at how existing non-blocking flushes work.

athanatos replied:
Probably in start_flush, if (!blocking) we should return an error.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9978,264 @@ struct C_Flush : public Context {
+  while (cur_off < oi.size && cur_off < max_window_size) {
+    bufferlist chunk_data;
+    // TODO: Do we need to support EC?
+    int r = pgbackend->objects_read_sync(
```
I guess we are assuming that we only trim a whole object at once, so if any part of the object is present, all of it is.  I presume that we return 0 immediately if the object has no non-dedup'd regions?  Where?

athanatos reviewed (changes requested):
Good PR!  I found the logic pretty easy to follow and I appreciate the separated commits.

Other than basic review comments, there are two things that need to be addressed for the next round:
1) I'm fine with disallowing non-blocking flushes for now, but make it clear early under what conditions it's an error.  After that, assert that the blocking conditions have happened.  This minimizes the possibility that we'll fail to clean up state somewhere.
2) Ensure that the behavior is sane if the object has been evicted.  This probably means returning sucess early (flushing an evicted object would have to be a noop) and asserting elsewhere that the object is present.

myoungwon commented:
@athanatos I addressed your comments. Can you take a look? Each commit since https://github.com/ceph/ceph/pull/37546/commits/518b78033aef57ca4b2354eb19e746cb3616a45d matches a comment you posted in order.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9982,280 @@ struct C_Flush : public Context {
+      info.oid =  p.first;
+      ctx->new_obs.oi.manifest.chunk_map[p.second.first] = info;
+      // drop all references issued before
+      refs.dec_ref(p.first);
```
I'm not sure I understand this.  Previously, I was unsure of what would happen if we flushed an object that already had some dedup'd sections.  I didn't see any logic to clear them from the chunk_map (maybe I missed it?).  Here, we have a new problem.  You appear to be clearing references on the *new* oids (mop->chunks reflects the chunks we are now writing out, right?).  That seems wrong.

Perhaps the answer is to revert to your previous version, but clear the chunk_map before adding the mop->chunks portions back in?

Prior to writing code, please respond to this comment with an explanation of your intention so that I can better understand what you are trying to do.

athanatos replied:
See my comment in start_dedup.

myoungwon replied:
I wrote the wrong code by mistake.

athanatos replied:
This probably means we're missing some test coverage as well.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9982,280 @@ struct C_Flush : public Context {
+  if (mop->results[0] < 0) {
+    // check if the previous op returns fail
+    if (mop->num_chunks == mop->results.size()) {
+      manifest_ops.erase(oid);
```
Is there a test for this case?  I think you need to propagate an error to the user here -- probably using       osd->reply_op_error as most existing error pathways do.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9982,280 @@ struct C_Flush : public Context {
+      chunk.substr_of(bl, p.first, p.second);
+      hobject_t target = get_fpoid_from_chunk(soid, chunk);
+      if (target == hobject_t()) {
+	dout(0) << " fingerprint oid is null " << dendl;
```
This is only possible if fp_algo is TYPE_FINGERPRINT_NONE, right?  I think you should check this case at the top and return -ENOTSUPP and assert here.

athanatos replied:
or EINVAL?

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9982,280 @@ struct C_Flush : public Context {
+	break;
+      }
+
+      // skip if the same content exits in prev snap at same offset
```
I mentioned in another comment that I'm not sure what's supposed to happen if the object already has some references.  Let's assume that the operation is:
1) calculate new fingerprint and chunks
2) drop existing references
3) take new references

I think you need to use object_manifest_t::calc_refs_to_drop_on_removal to calculate the references you need to drop.  You can then use calf_refs_to_inc_on_set to get the set of references you need to add.  Those already correctly handle snapshots and should simplify your code.  Note, you need to defer dropping the references until after commit.

athanatos replied:
If this doesn't match your intentions, let's talk about it here.

myoungwon replied:
Sorry, my bad.

athanatos reviewed (changes requested):
Getting closer, main thing left to address is handling of snapshots and existing references.

myoungwon commented:
@athanatos Based on your comment, I've added commits to correct a test case and handle of the existing reference for snapshots.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9982,312 @@ struct C_Flush : public Context {
+    return 0;
+  }
+  if (pool.info.get_fingerprint_type() == pg_pool_t::TYPE_FINGERPRINT_NONE) {
+    dout(0) << " fingerprint algorithm is not set " << dendl;
```
Why would this be success?  I think this should return ENOTSUPP or EINVAL to the user.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9982,312 @@ struct C_Flush : public Context {
+      // all operations are finished
+      mop->new_chunk_map[p.first] = chunk_info;
+      // skip if the same content exits in prev snap at same offset
+      if (obc->ssc->snapset.clones.size()) {
```
This portion below is unnecessarily wasteful.  You call get_prev_clone_obc for every chunk when clearly get_prev_clone_obc is a function of the object and not of the chunk.  You could easily declare set_chunk before the loop and then call calc_refs_to_inc_on_set after the loop on the accumulated set of references with cobc after calling get_prev_clone_obc a single time.  Note, calc_refs_to_inc_on_set *specifically* tolerates a null value passed in for both _g an _l, you shouldn't need to guard this call with an if.

athanatos replied:
You haven't addressed this.  As I mentioned above, the outer while loop doesn't appear to be useful as bl will always contain the entire object after you call do_cdc.  
* Move get_fpoid_from_chunk into do_cdc and have it populate a mop->new_chunk_map directly.
* Change the type of mop->new_chunk_map to an object_manifest_t named mop->new_manifest.
* Once it is fully populated, pass mop->new_manifest to calc_refs_to_inc_on_set *once* and then use the resulting map to issue the refcounts.

I'd also really prefer that you not duplicate the logic in refcount_manifest.  At  least, I only want once place that encodes the "chunk_get_ref" names.  You should be able to modify refcount_manifest or one of the higher level helpers enough to be useful here.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9982,312 @@ struct C_Flush : public Context {
+	  object_manifest_t set_chunk;
+	  set_chunk.chunk_map[p.first] = chunk_info;
+	  set_chunk.calc_refs_to_inc_on_set(
+	    &cobc->obs.oi.manifest,
```
You are passing in the the prev clone as _g rather than _l.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9982,312 @@ struct C_Flush : public Context {
+	    continue;
+	  } 
+	}
+      }
```
This implicit else is wrong -- if there isn't a cobc, you still need to calc_refs_to_inc_on_set.

athanatos replied:
I see, you are actually ignoring refs.  The whole point of object_ref_delta_t and calc_refs_to_inc_on_set is that you build up an object_manifest_t of the refs you need to set, call calc-refs_to_inc_on_set with either null or non-null clones depending on whether they exist, and then use the resulting refs to send the chunk operations.

athanatos replied:
You probably ought to use get_adjacent_clones as below to appropriately the clone case.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9982,312 @@ struct C_Flush : public Context {
+    * So, the precedure is 
+    * 	1. calc_refs_to_drop_on_removal()
+    * 	2. drop old references by dec_refcount()
+    * 	3. update new chunk_map
```
This comment is misleading.  2. here happens *after* 3. (you are using register_on_commit to call dec_refcount).  It must be that way because we cannot drop the references to the old set until we've committed the new set.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9982,312 @@ struct C_Flush : public Context {
+      return;    
+    }
+
+    object_ref_delta_t refs;
```
Move this declaration to immediately above calc_refs_to_drop_on_removal.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9982,312 @@ struct C_Flush : public Context {
+    * 20:   [0, 2) ddd, [6, 2) bbb, [8, 2) ccc
+    * 
+    * In this case, if the new chunk_map is as follows,
+    * new_chunk_map : [0, 2) DDD, [6, 2) bbb, [8, 2) ccc
```
DDD -> ddd to match the rest of the comment.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9982,312 @@ struct C_Flush : public Context {
+    get_adjacent_clones(ctx->obs->oi, ctx.get(), obc_l, obc_g);
+    // clear all old references
+    ctx->obs->oi.manifest.calc_refs_to_drop_on_removal(
+      obc_l ? &(obc_l->obs.oi.manifest) : nullptr,
```
This logic appears to correctly handle the case where the object is a clone -- start_dedup probably should as well.

myoungwon commented:
@athanatos Can you take a look? I addressed your comments.

athanatos commented:
I'm off the rest of this week, I'll take a look on Monday.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +9982,312 @@ struct C_Flush : public Context {
+  uint64_t cur_off = 0;
+  string chunk_algo = pool.info.get_dedup_chunk_algorithm_name();
+  int64_t chunk_size = pool.info.get_dedup_cdc_chunk_size();
+  uint64_t max_window_size = static_cast<uint64_t>(pool.info.get_dedup_cdc_window_size());
```
Is this really only meant to bound how much of the object we read at once?  This method is going to read the *whole object* into memory anyway.  The while loop in start_dedup doesn't appear to make sense, this method will always read the whole object anyway.  Moreover, there's no reason to pass in a bufferlist reference here -- the only reason you do is so that start_dedup can then then run get_fpoid_from_chunk.  Why not just do that here and return the chunk_map directly?

You need to carefully rethink how do_cdc and start_dedup work.  If you are having trouble with it, I can write a cleaner version later this week or next.

myoungwon commented:
@athanatos I added two commits based on your comments. Can you check the commits?

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +10022,278 @@ struct C_Flush : public Context {
+  } 
+
+  if (mop->tids.size() == 0) {
+    return 0;
```
Combine these two ifs, and put the other return in an else.

athanatos commented:
I'll try to have a look before the end of the week.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +10022,278 @@ struct C_Flush : public Context {
+    chunk_map[object_offset] = chunk_info_t(0, p.second, target);
+    total_length += p.second;
+  }
+  return total_length;
```
Let's say the window size is 1MB and do_cdc is configured to emit chunks of size 16k - 32k.  If the second to last chunk is 1MB - 20k, the last chunk might not be on an actual chunk boundary.  calc_chunks is going to return it as a chunk anyway because it is the end of the buffer, but on a 4MB object, it would tend to prevent dup detection with other examples of the same string.  You could resolve this by rewinding total_length to the offset of the end of the last whole chunk, but for now I'd rather you read in the whole object here as you did before (but with a single call to object_read_sync),  call do_cdc on the whole thing, eliminate the loop in start_dedup, and eliminate the pool.info.get_dedup_cdc_window_size() configuration parameter entirely.

myoungwon commented:
@athanatos Done.

athanatos commented:
I'll try to have another look on Monday.  It's probably worth the time for you to take another pass over it and ensure that everything is as clear and simple as possible.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +10022,267 @@ struct C_Flush : public Context {
 
+int PrimaryLogPG::start_dedup(OpRequestRef op, ObjectContextRef obc)
+{
+  bufferlist bl;
```
Unused?

myoungwon replied:
Fixed.

athanatos commented:
C_SetManifestRefCountDone leaks cb on r == -ECANCELED.  cb should be set to null after the call to cb->complete(r) and ~C_SetManifestRefCountDone should delete cb.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -3395,38 +3420,51 @@ void PrimaryLogPG::dec_all_refcount_manifest(const object_info_t& oi, OpContext*
-      new C_SetManifestRefCountDone(cb, src_soid);
+    if (type == refcount_t::INCREMENT_REF ||
+        type == refcount_t::DECREMENT_REF) {
+      fin = new C_SetManifestRefCountDone(static_cast<RefCountCallback*>(cb), src_soid);
```
This is a really confusing asymmetry for this helper.  You needed to do this because start_dedup issues multiple refcount_manifest calls on a single manifest_ops entry, so you need to assume that the caller has already handled manifest_ops and cb already takes care of it. It just so happens that start_dedup is the only user of CREATE_OR_GET_REF, so special casing it avoided the problem at the expense of making this helper's behavior randomly different between CREATE_OR_GET_REF and INCREMENT_REF.  DECREMENT_REF never passes cb, so that case is irrelevant here.  INCREMENT_REF happens in two places, I think you should should move the manifest_ops handling inline in those two places or create a helper to do so and make this method agnostic of what cb does.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +10022,266 @@ struct C_Flush : public Context {
+            << " len: " << oi.size << " r: " << r << dendl;
+    return r;
+  }
+  if (bl.length() == 0) {
```
bl.length() != oi.size

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +10022,266 @@ struct C_Flush : public Context {
+    } else if (mop->op) {
+      dout(10) << __func__ << " waiting on write lock " << mop->op << dendl;
+      close_op_ctx(ctx.release());
+      osd->reply_op_error(mop->op, -EAGAIN);
```
No, being unable to get a lock does not mean we can reply with an error to the client.  get_lock_type() returning false here means that mop->op got placed on a wait list.  You should call close_op_ctx and *return* -EAGAIN up to do_osd_ops.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -9976,6 +10022,266 @@ struct C_Flush : public Context {
+    }
+
+    // clear all regions
+    ctx->new_obs.oi.manifest.chunk_map.clear();
```
This is unnecessary with the operator= usage immediately below, right?

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -10124,9 +10124,10 @@ int PrimaryLogPG::do_cdc(const object_info_t& oi,
+  if (bl.length() != oi.size) {
+    dout(0) << __func__ << " bl.length: " << bl.length() << " != oi.size: "
+	    << oi.size << " during chunking " << dendl;
+    return -ENODATA;
```
Probably -EIO or something, this would mean that the oi is inconsistent with the ondisk state.

myoungwon commented:
@athanatos Updated. Other comments?

athanatos commented:
Ok, I think it's ready for teuthology testing.

myoungwon commented:
@athanatos 
https://pulpito.ceph.com/myoungwon-2020-12-03_14:50:10-rados-wip-revamp-tier-flush-distro-basic-smithi/
https://pulpito.ceph.com/myoungwon-2020-12-04_03:52:15-rados-wip-revamp-tier-flush-distro-basic-smithi/

It seems that the failure is not related to this PR.
The only failure has been already tracked by https://tracker.ceph.com/issues/48323.
I posted the pull request to resolve the issue (https://github.com/ceph/ceph/pull/38237).
After merging this PR, I will focus on that PR.

athanatos commented:
rerun tests please

athanatos commented:
jenkins test make check

athanatos commented:
jenkins retest this please
